spec/index.bs

```diff
@@ -2793,17 +2797,44 @@ GPUQuerySet includes GPUObjectBase;
 dictionary GPUQuerySetDescriptor : GPUObjectDescriptorBase {
     required GPUQueryType type;
     required GPUSize32 count;
+    GPUPipelineStatisticFlags pipelineStatistics;
 };
 </script>
 
+  * {{GPUQuerySetDescriptor/pipelineStatistics}}: a bitmask of {{GPUPipelineStatisticFlags}} specifying which pipeline statistics will be returned in the new query set.
+    |pipelineStatistics| is ignored if type is not {{GPUQueryType/pipeline-statistics}}.
+    <div class=validusage dfn-for=GPUQuerySetDescriptor.pipelineStatistics>
+      <dfn abstract-op>Valid Usage</dfn>
+        1. If {{GPUExtensionName/pipeline-statistics-query}} is not available, |type| must not be {{GPUQueryType/pipeline-statistics}}.
+        2. If |type| is {{GPUQueryType/pipeline-statistics}},  |pipelineStatistics| must be a valid combination of {{GPUPipelineStatisticFlags}} values.
+    </div>
+
 ## QueryType ## {#querytype}
 
 <script type=idl>
 enum GPUQueryType {
-    "occlusion"
+    "occlusion",
+    "pipeline-statistics"
+};
+</script>
+
+## Pipeline Statistics Query ## {#pipeline-statistics}
+
+<script type=idl>
+typedef [EnforceRange] unsigned long GPUPipelineStatisticFlags;
+interface GPUPipelineStatisticBit {
+    const GPUPipelineStatisticFlags VERTEX_SHADER_INVOCATIONS   = 0x01;
+    const GPUPipelineStatisticFlags CLIPPER_INVOCATIONS         = 0x02;
+    const GPUPipelineStatisticFlags CLIPPER_PRIMITIVES_OUT      = 0x04;
+    const GPUPipelineStatisticFlags FRAGMENT_SHADER_INVOCATIONS = 0x08;
+    const GPUPipelineStatisticFlags COMPUTE_SHADER_INVOCATIONS  = 0x10;
 };
 </script>
 
+When resolving the results of a pipeline statistics query, the number of pipeline statistics written into GPU buffer depends on the number of enabled {{GPUPipelineStatisticBit}} in {{GPUQuerySetDescriptor/pipelineStatistics}}.
+
+Each result is written into unsigned 64-bit integer and in the order of the lowest-valued member of the {{GPUPipelineStatisticBit}} to the highest.
```

Comment:
Are we relying on the backends for this behavior? If so, are they consistent, or is fixup required?

Reply:
This behavior follows Vulkan rules, and need fixup on D3D12 and Metal, they will query **all** of pipeline statistics and resolve results into buffer according to the structure of  [D3D12_QUERY_DATA_PIPELINE_STATISTICS ](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_query_data_pipeline_statistics) or [MTLCounterResultStatistic](https://developer.apple.com/documentation/metal/mtlcounterresultstatistic?language=objc), we need to pick  the values corresponding to the set bits and write them to buffer in order.